core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```
@@ -0,0 +1,155 @@
+    }
+    writeToFile(matchedLines);
+  }
+
+  private void listFilesRecursive(File dir) {
+    if (!dir.exists()) {
+      return;
+    }
+    if (dir.isDirectory()) {
+      for (File file : Objects.requireNonNull(dir.listFiles())) {
+        listFilesRecursive(file);
+      }
+    } else {
+      this.files.add(dir);
+    }
+  }
+
+  @Override
+  public List<File> listFiles(String rootDir) {
+    this.files = new ArrayList<>();
+    File dir = new File(rootDir);
+
+    listFilesRecursive(dir);
+    return this.files;
+  }
+
+  @Override
+  public Stream<String> readLines(File inputFile) {
+    List<String> lines = new ArrayList<>();
+
+    try {
+      String line;
+      BufferedReader reader = new BufferedReader(new FileReader(inputFile));
+      while ((line = reader.readLine()) != null) {
+        lines.add(inputFile + ":" + line);
+      }
+      reader.close();
+    } catch (FileNotFoundException ex) {
+      this.logger.error(ex.getMessage(), ex);
+    } catch (IOException ex) {
+      this.logger.error(ex.getMessage(), ex);
+    }
+
+    return lines.stream();
+  }
+
+  @Override
+  public boolean containsPattern(String line) {
+    return line.matches(this.regex);
+  }
+
+  @Override
+  public void writeToFile(List<String> lines) throws IOException {
+    BufferedWriter bufferedWriter = new BufferedWriter(
+        new OutputStreamWriter(new FileOutputStream(this.outFile)));
+
+    for (String line : lines) {
+      bufferedWriter.write(line);
+      bufferedWriter.newLine();
+    }
+    bufferedWriter.close();
+  }
+
+  public static void main(String[] args) {
+    if (args.length != 3) {
+      throw new IllegalArgumentException("");
+    }
+
+    //Use default logger config
+    BasicConfigurator.configure();
+
+    JavaGrepImp javaGrepImp = new JavaGrepImp();
+    javaGrepImp.setRegex(args[0]);
+    javaGrepImp.setRootPath(args[1]);
+    javaGrepImp.setOutFile(args[2]);
+
+    try {
+      javaGrepImp.process();
+    } catch (IOException ex) {
```

you want to catch all exception `Exception ex`